core/src/value/deserialize.rs

```diff
@@ -29,55 +50,286 @@ pub enum Error {
+    #[snafu(display(
+        "Could not parse incomplete value: first missing component: {:?}",
+        component
+    ))]
+    IncompleteValue {
+        component: DateComponent,
+        backtrace: Backtrace,
+    },
+    #[snafu(display("Range is zero"))]
+    ZeroRange { backtrace: Backtrace },
 }
 
 type Result<T, E = Error> = std::result::Result<T, E>;
 
+#[derive(Debug, PartialEq)]
+pub enum DateComponent {
+    Year,
+    Month,
+    Day,
+    Hour,
+    Minute,
+    Second,
+    Fraction,
+    UTCOffset,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum DatePrecision {
+    Year(u16),
+    Month(u16, u8),
+    Day(u16, u8, u8),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PartialDate {
+    components: DatePrecision,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum DateTimePrecision {
+    Year(u16),
+    Month(u16, u8),
+    Day(u16, u8, u8),
+    Hour(u16, u8, u8, u8),
+    Minute(u16, u8, u8, u8, u8),
+    Second(u16, u8, u8, u8, u8, u8),
+    Fraction(u16, u8, u8, u8, u8, u8, u32),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PartialDateTime {
+    components: DateTimePrecision,
+    // TODO probably not an option, will see ..
+    offset: Option<FixedOffset>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum TimePrecision {
+    Hour(u8),
+    Minute(u8, u8),
+    Second(u8, u8, u8),
+    Fraction(u8, u8, u8, u32),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PartialTime {
+    components: TimePrecision,
+}
+
+/**
+ * This trait is implemented by partial precision
+ * Date, Time and DateTime structures.
+ * Trait method returns the last fully precise component of the value.
+ */
+pub trait Precision {
+    fn precision(&self) -> DateComponent;
+}
+
+impl Precision for PartialDate {
+    fn precision(&self) -> DateComponent {
+        match self.components {
+            DatePrecision::Year(..) => DateComponent::Year,
+            DatePrecision::Month(..) => DateComponent::Month,
+            DatePrecision::Day(..) => DateComponent::Day,
+        }
+    }
+}
+
+/**
+ * This trait is implemented by partial precision
+ * Date, Time and DateTime structures.
+ */
+pub trait Imprecision<U>
+where
+    Self: Precision,
+    U: PartialEq,
+{
+    /**
+     * Returns the earliest possible value from a partial precision structure.
+     * So missing components default to 1 (days, months) or 0 (hours, minutes, ...)
+     * If structures contain invalid combination of date/time components, it fails.
+     */
+    fn earliest(&self) -> Result<U>;
+
+    /**
+     * Returns the latest possible value from a partial precision structure.
+     * If structures contain invalid combination of date/time components, it fails.
+     */
+    fn latest(&self) -> Result<U>;
+
+    /**
+     * Returns a tuple of the earliest and latest possible value from a partial precision structure.
+     * 
+     */
+    fn to_range(&self) -> Result<(Option<U>, Option<U>)> {
+        if self.is_precise() {
+            ZeroRange.fail()
```

Comment:
This will depend a bit on how `is_precised` will be designed, but if the range extremities are inclusive, creating a range should succeed even if a single specific date-time is given. For `U = PartialDate`, for example, `2021-08-28` would return a range that matches the date itself.